### ISSUE TITLE
Magic Paper

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#2409]
Magic Paper<br>
<br>
http://bit.ly/2h0Ptnk<br>
<br>
<br>
<br>
<div>
<div>
<h1>Magic Paper</h1>
<p>Working notes by <a href="http://michaelnielsen.org"><em>Michael Nielsen</em></a>, November 2017. Followup to (but doesn't require) my <a href="http://cognitivemedium.com/interfaces-1/index.html">notes on Chalktalk</a>.</p>
<blockquote>
<p><em>[Technology] transforms what can be said and what is said&hellip; all major advances in consciousness depend on technological transformations in implementations of the word.</em> &ndash; Walter Ong, <em>Interfaces of the Word</em> (1977).</p>
</blockquote>
<p>Mathematical ideas are conventionally expressed using notation and terminology developed using static media. Suppose, however, that mathematics had been invented <em>after</em> modern computers. This is perhaps difficult to imagine &ndash; after all, mathematics helped lead to computers &ndash; but let's do the thought experiment anyway. Might mathematical notation have developed in a different way? Would we instead have developed a dynamic, interactive notation more powerful than the static mathematical and linguistic notations in common use today?</p>
<p>As part of my exploration of this question, I recently developed a simple prototype system, which I call <em>Magic Paper</em> (<a href="https://github.com/mnielsen/magic_paper">code</a>). To get a feel for what the system does, below is a video of me using <em>Magic Paper</em> to explain the (very simple!) basic idea underlying the proof of a beautiful theorem about artificial neural networks: the result that neural nets may be used to approximate any continuous function*<span>* See <a href="http://cognitivemedium.com/magic_paper/assets/Cybenko.pdf">Approximation by superpositions of sigmoidal function</a>, by George Cybenko (1989), and <a href="http://cognitivemedium.com/magic_paper/assets/Hornik.pdf">Multilayer feedforward networks are universal approximators</a>, by Kurt Hornik, Maxwell Stinchcombe, and Halbert White (1989).</span>. Note that you do not need to be deeply familiar with neural networks to follow along; an introduction to the basics of neural nets may be found <a href="http://neuralnetworksanddeeplearning.com/chap1.html">here</a>. You may wish to fullscreen the video, for easier viewing:</p>
VIDEO
<p>The design of <em>Magic Paper</em> was inspired in part by a system developed by Ken Perlin, known as <em>Chalktalk</em>. <em>Chalktalk</em> has not been publicly released at the time of this writing (November 2017)*<span>* <strong>Update:</strong> Chalktalk has since been released, see <a href="https://github.com/kenperlin/chalktalk">here</a>.</span>, but you can get some understanding of the system by viewing from 32:40 to the end of the following video:</p>
VIDEO
<h3>Can a visual medium be used to do serious mathematical exploration?</h3>
<p>I built <em>Magic Paper</em> to explore two questions not directly addressed by <em>Chalktalk</em>.</p>
<p>The first question is whether such a visual medium can be used to do serious mathematical exploration, the kind a researcher might do when working on a research problem. In other words: can such a system be used as a medium for creative thought? While <em>Chalktalk</em> is a fun medium to explore in, it's not obvious to me that it's suitable for the kind of exploration that arises in mathematical work.</p>
<p>Of course, it's too ambitious for such a quickly-built prototype medium to be useable across all of mathematics. And so I decided to narrow the question, and ask whether such a medium could be used to explore at least <em>one</em> non-trivial mathematical question; concretely: can neural nets be used to approximate any function?</p>
<p>While focusing on one problem is much easier than building a general-purpose exploratory mathematical medium, it's still an extremely demanding and thus informative constraint. Furthermore, it seems likely that patterns useful in solving one mathematical problem will be reuseable in solving others.</p>
<p>In fact, even this is too demanding. The idea for a proof presented above wasn't actually discovered in this medium. It's a proof taken from my book on neural networks*<span>* Michael A. Nielsen, <a href="http://neuralnetworksanddeeplearning.com/chap4.html">Neural Networks and Deep Learning (Chapter 4)</a>, Determination Press (2015).</span> . Despite this, I believe this context &ndash; explaining the idea underlying the proof of a relatively recent mathematical result &ndash; is a good one for understanding what is required for a general-purpose exploratory mathematical medium.</p>
<p>Note that while the video of <em>Magic Paper</em> perhaps makes it appear to be explanatory or educational in intent, that is not the case. It is intended primarily to be a medium to think and explore in, and only secondarily to explain or to educate in. While these goals are sometimes congruent, they are also often in tension.</p>
<p>As an example of this tension, explanatory media are often highly polished, presenting a finished product from which distracting elements have been eliminated. By contrast, exploration involves distractions, wrong paths, and dead ends. One sees this in myriad small ways in the video: the frequent readjustment of the position and size of the glyphs; the various unnecessary weight and bias parameters that are sometimes shown; and so on. In a polished presentation these would have been eliminated, showing just what the viewer needs at any given moment. But such superfluous elements are inevitable in exploration.</p>
<p>Another example of the tension between education and exploration comes from the programming language Logo. Logo is often touted as an example of discovery-based learning, and in many respects it's a good exploratory medium for mathematics. However, while Logo has had considerable success in helping children understand ideas about differential geometry, differential geometers do not, as far as I'm aware, use it as a medium for research. That's potentially a problem: when education and real practice are separated, it can result in busy work on poorly-motivated &ldquo;educational&rdquo; problems. An interesting project would be to create a single environment which can be used for both education and research. I am optimistic this is possible; after all, beginning chess players use the same board and the same moves as the world champion.</p>
<h3>Can we reify deep ideas about mathematics inside the <em>Magic Paper</em> interface?</h3>
<p>The second question motivating <em>Magic Paper</em> was whether it is possible to reify deep ideas about mathematics inside the interface. That sounds cryptic, but has a straightforward meaning. To explain that meaning, let's briefly switch topic, and recall an idea from elementary physics, the principle of conservation of energy:</p>
<center>
<p><em>&frac12;mv<sup>2</sup> + U(x)</em> = constant</p>
</center>
<p>The basic point is that there's some relationship between the mass of a particle, <em>m</em>, its velocity <em>v</em>, its position <em>x</em>, and its potential energy <em>U(x)</em>. The way physics is ordinarily done, we manipulate this expression using algebraic operations. But a different approach is possible. In my essay &ldquo;Thought as a Technology&rdquo;*<span>* Michael Nielsen, <a href="http://cognitivemedium.com/tat/index.html">Thought as a Technology</a>, available at http://bit.ly/2fK70A6 (2016).</span> I develop a prototype interface which reifies the principle of conservation of energy. In particular, the prototype vividly shows the user the relationship between mass, position, velocity and potential energy. This makes it possible to directly see and adjust the potential energy, see how the trajectories respond, and so on. This gives a unique and rich experience, one enabling the user to internalize a powerful set of cognitive technologies for thinking about motion.</p>
<p><em>Chalktalk</em>-like media seem promising for reifying such principles. Unfortunately, as far as I'm aware there are no general results about neural networks as powerful as conservation of energy. However, as a step in this direction, directly seeing the relationship between the parameters of a network and the output of various neurons does help us understand the behaviour of the network. Further work could incorporate deeper ideas about neural networks, such as learning by gradient descent, or regularization, or ways of visualizing the network*<span>* A nice recent example is the paper by Jason Yosinksi, Jeff Clune, Anh Nguyen, Thomas Fuchs, and Hod Lipson, on <a href="http://cognitivemedium.com/magic_paper/assets/Yosinski2015a.pdf">Understanding Neural Networks Through Deep Visualization</a> ICML (2015).</span>. It'd also be interesting to develop such an interface in a context where more deep principles are available &ndash; perhaps using <em>Magic Paper</em> to explore (say) chaotic dynamical systems, using ideas such as conservation of energy, conservation of phase space volume, Lyapunov exponents, and the KAM theorem. Building such ideas into the interface would make them a routine, concrete part of the way the user works and thinks.</p>
<h3>Why not just key commands?</h3>
<p>Instead of drawing a neuron, why not just hit &ldquo;n&rdquo;? Instead of drawing axes, why not just hit &ldquo;g&rdquo;, for &ldquo;graph&rdquo;? Wouldn't such keystrokes be faster and easier?</p>
<p>This question has bothered me ever since first seeing Perlin demonstrate <em>Chalktalk</em>. Even as an audience member, I found it much more satisfying to see him draw a glyph than to just hit a key and have it appear. And this effect was even stronger as a user of <em>Magic Paper</em>.</p>
<p>I don't have a satisfactory understanding of why it's satisfying to do this. Perhaps it's because the act of drawing activates associations in my brain that wouldn't have been activated as strongly by issuing a key command. In this sense, the act of drawing is far more meaningful &ndash; and meaning-inducing &ndash; than hitting a key. It's a little like the reasons we don't name a variable meant to represent someone's name as <em>jnwhg</em>; it's more meaningful to use <em>name</em>. Or the reasons it's better to use a map of a coastline to navigate, rather than a log of latitudes and longitudes for points on the coast.</p>
<p>Unfortunately, while such explanations are suggestive they're vague and incomplete. And yet my lived experience is that this is a strong effect. Something similar is true of physical activities, such as tennis or dance: there is an enormous difference between a symbolic representation &ndash; say, reading about tennis &ndash; versus actually doing it. Now, for the most part, we tend to think of these activities as a special class of &ldquo;physical&rdquo; activities, different from &ldquo;mental&rdquo; activities such as mathematics. Yet I'm not sure the distinction is all that sharp.</p>
<p>More broadly, advocates of embodied cognition often make strong claims about its importance. As far as I can see, the evidence they have for those claims is weak. And yet while the specific claims may be weak or wrong, I believe there is some strong and important underlying effect. It really does greatly help my thinking to sketch the glyphs in <em>Magic Paper</em>. I just don't understand why.</p>
<h3>Working within the system</h3>
<p>I began the design work for <em>Magic Paper</em> with paper-and-pen storyboards and a text-based script for my video. As I developed the storyboards and script I also began to develop the system. Very quickly, the system was powerful enough that I could do basic sketches. Once that point was reached, it was more stimulating (and easier) to work within the new medium, and to abandon the storyboards and script.</p>
<p>I initially felt guilty at abandoning the storyboards and script, as though that was somehow the &ldquo;correct&rdquo; way of working. But working in the new medium quickly began to suggest useful ideas that I do not believe I would otherwise have anticipated.</p>
<p>As an example, in the final video we often see multiple graphs showing the behaviour of several different neurons in a network. This makes it easy for the viewer to understand the behaviours of different parts of the network, and to put it all together:</p>
<center><img src="http://cognitivemedium.com/magic_paper/assets/multiple_graphs.png"></center>
<p>By contrast, my initial storyboards and script had just a single output graph. It was only after I began using the medium that I realized it was possible to use the graphs in this way. In particular, I didn't design the system so it could plot the behavior of intermediate neurons. Rather, that was a fortuitous happenstance, tried spontaneously as I used the medium.</p>
<p>Another point I did not anticipate in storyboarding and scripting was how useful it would be to move and resize the neurons and the graphs, e.g.:</p>
<center><img src="http://cognitivemedium.com/magic_paper/assets/movement_edited.gif"></center>
<p>A big benefit of this is object persistence. When possible, it's often useful to reuse or repurpose existing objects, since viewers have already built strong mental models of their behaviour. This kind of dynamic reuse is harder to achieve in static media. Perhaps for that reason I did not anticipate it in my storyboarding and scripting.</p>
<p>By contrast, my earlier description of this proof in an online book consisted of text interspersed with interactive elements, e.g.:</p>
<img src="http://cognitivemedium.com/magic_paper/assets/nnadl_4_pt1.png"><img src="http://cognitivemedium.com/magic_paper/assets/nnadl_4_pt2.png">
<p>Many of the interactive elements depict minor variations on the same system. But because there is no object persistence, the user must parse each one separately, and put in the mental work required to understand the relationships. For the most part this is not terribly hard work, but it is an additional mental burden.</p>
<p>All told, while the script and storyboards were useful to bootstrap, once the medium was running, it was more generative to work directly within that, experimenting and evolving an approach I was comfortable with.</p>
<h3>Extensions: abstracting glyphs, re-parameterization, glyph design</h3>
<p>Here are three extensions that could be added to <em>Magic Paper</em> with 40 to 50 hours of additional work.</p>
<p>First, it should be possible to &ldquo;lasso&rdquo; several elements, combining them into a single new glyph. For instance, we could lasso the two-neuron gadget used to build tower functions, and turn it into a single re-usable glyph. That abstracted glyph could then be duplicated, moved, its parameters edited, and so on, just as any other glyph. It would also be possible to inspect the internal state of the abstracted glyph, making it possible to understand multiple layers of abstraction.</p>
<p>Second, it should be possible to re-parameterize the abstracted glyph, so it has just three parameters, representing the height of the tower function, a starting <em>x</em> value, and a finishing <em>x</em> value. These three parameters would then determine the biases and weights of the underlying elements. This could be implemented with a simple expression parser.</p>
<p>Third, it should be possible to integrate an image editor enabling the addition of extra glyphs to <em>Magic Paper</em>'s visual language. Amongst other things, such an editor could be used to add unique visual representations for the abstracted glyphs.</p>
<h3>Conclusion</h3>
<p>My research colleagues sometimes express puzzlement at my interest in new media forms. They often treat it as fun but not especially serious work. To them, advanced research usually means work on esoteric subjects. But people working on new media forms have the opportunity to rethink absolutely fundamental concepts such as number, equality, and so on; such work is about rethinking what mathematics and language is. I believe such experiments will eventually affect and perhaps transform the way all research is done. But it will take a huge effort, since this kind of rethinking is not an individual-sized problem, it is a civilization-sized problem, and we're just getting started.</p>
<h2>Acknowledgements</h2>
<p>Thanks to Shan Carter, Nicky Case, Andy Matuschak, Robert Ochshorn, Kanjun Qiu, Grant Sanderson, and Jack Schaedler for helpful discussions.</p>
<h2>Citation</h2>
<p>In academic work, please cite this as: Michael Nielsen, <a href="http://cognitivemedium.com/magic_paper/index.html">&ldquo;Magic Paper&rdquo;</a>, available at http://bit.ly/2h0Ptnk (2017).</p>
<h2>Further Reading</h2>
<p>Michael Nielsen, <a href="http://cognitivemedium.com/emm/emm.html">Toward an Exploratory Medium for Mathematics</a> (2016).</p>
<p>Michael Nielsen, <a href="http://cognitivemedium.com/tat/index.html">Thought as a Technology</a> (2016).</p>
<p>Bret Victor, <a href="http://worrydream.com/KillMath/">Kill Math</a> (2011). See also my <a href="http://mnielsen.github.io/notes/kill_math/kill_math.html">notes on Kill Math</a> (2014).</p>
<p>Bret Victor, <a href="http://worrydream.com/MediaForThinkingTheUnthinkable/">Media for Thinking the Unthinkable</a> (2013).</p>
<p>Bret Victor, <a href="http://worrydream.com/MagicInk/">Magic Ink</a> (2006).</p>
<p>Seymour Papert, <a href="https://mindstorms.media.mit.edu/">Mindstorms: children, computers, and powerful ideas</a> (1980).</p>
<p>Jason Brennan, Scott Farrar, Natalie Fitzgerald, May-Li Khoe, and Andy Matuschak, <a href="https://www.khanacademy.org/research/reports/early-math">Playful worlds of creative math: a design exploration</a> (2017).</p>
<h2>Licensing</h2>
<p>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc/3.0/deed.en_GB">Creative Commons Attribution-NonCommercial 3.0 Unported License</a> . This means you're free to copy, share, and build on it, but not to sell it. If you're interested in commercial use, please <a href="mailto:mn@michaelnielsen.org">contact me</a>. <a rel="license" href="http://creativecommons.org/licenses/by-nc/3.0/deed.en_GB"><img alt="Creative Commons Licence" src="http://i.creativecommons.org/l/by-nc/3.0/88x31.png"></a></p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Cognitive Medium<br>
<br>
January 13, 2019 at 04:14PM
